compiler/rustc_parse_format/src/lib.rs

```diff
@@ -809,94 +838,97 @@ impl<'a> Parser<'a> {
 /// Finds the indices of all characters that have been processed and differ between the actual
 /// written code (code snippet) and the `InternedString` that gets processed in the `Parser`
 /// in order to properly synthesise the intra-string `Span`s for error diagnostics.
-fn find_skips_from_snippet(
+fn find_width_map_from_snippet(
     snippet: Option<string::String>,
     str_style: Option<usize>,
-) -> (Vec<usize>, bool) {
+) -> (Vec<InnerWidthMapping>, bool) {
     let snippet = match snippet {
         Some(ref s) if s.starts_with('"') || s.starts_with("r\"") || s.starts_with("r#") => s,
         _ => return (vec![], false),
     };
 
-    fn find_skips(snippet: &str, is_raw: bool) -> Vec<usize> {
+    fn find_width_map(snippet: &str, is_raw: bool) -> Vec<InnerWidthMapping> {
         let mut s = snippet.char_indices().peekable();
-        let mut skips = vec![];
+        let mut width_mappings = vec![];
         while let Some((pos, c)) = s.next() {
             match (c, s.peek()) {
                 // skip whitespace and empty lines ending in '\\'
-                ('\\', Some((next_pos, '\n'))) if !is_raw => {
-                    skips.push(pos);
-                    skips.push(*next_pos);
-                    let _ = s.next();
+                ('\\', Some((_, '\n'))) if !is_raw => {
+                    let mut width = 2;
+                    s.next();
 
-                    while let Some((pos, c)) = s.peek() {
+                    while let Some((_pos, c)) = s.peek() {
                         if matches!(c, ' ' | '\n' | '\t') {
-                            skips.push(*pos);
-                            let _ = s.next();
+                            width += 1;
+                            s.next();
                         } else {
                             break;
                         }
                     }
+
+                    width_mappings.push(InnerWidthMapping::new(pos, width, 0));
                 }
-                ('\\', Some((next_pos, 'n' | 't' | 'r' | '0' | '\\' | '\'' | '\"'))) => {
-                    skips.push(*next_pos);
-                    let _ = s.next();
+                ('\\', Some((_, 'n' | 't' | 'r' | '0' | '\\' | '\'' | '\"'))) => {
+                    width_mappings.push(InnerWidthMapping::new(pos, 2, 1));
+                    s.next();
                 }
                 ('\\', Some((_, 'x'))) if !is_raw => {
+                    // consume `\xAB` literal
                     for _ in 0..3 {
-                        // consume `\xAB` literal
-                        if let Some((pos, _)) = s.next() {
-                            skips.push(pos);
-                        } else {
-                            break;
-                        }
```

Comment:
Note: a small check removed here. The lexer will raise "numeric character escape is too short" before we ever get to this code if there aren't three characters following the `\`.